Mathlib/NumberTheory/EulerProduct/Basic.lean

```diff
@@ -159,37 +329,55 @@ lemma summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric {f : ℕ →*
+  exact summable_and_hasSum_factoredNumbers_prod_filter_prime_geometric h _
 
 /-- A version of `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric`
 in terms of the value of the series. -/
 lemma prod_primesBelow_geometric_eq_tsum_smoothNumbers {f : ℕ →* F} (hsum : Summable f) (N : ℕ) :
     ∏ p in N.primesBelow, (1 - f p)⁻¹ = ∑' m : N.smoothNumbers, f m := by
-  refine (summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric ?_ N).2.tsum_eq.symm
-  exact fun {_} hp ↦ hsum.norm_lt_one hp.one_lt
+  rw [smoothNumbers_eq_factoredNumbers, primesBelow]
+  exact prod_filter_prime_geometric_eq_tsum_factoredNumbers hsum _
+
+/-- The *Euler Product* for completely multiplicative functions.
+
+If `f : ℕ →*₀ F`, where `F` is a complete normed field and `‖f ·‖` is summable, then
+`∏' p : Nat.Primes, (1 - f p)⁻¹ = ∑' n, f n`.
+This version is stated in terms of `HasProd`. -/
+theorem eulerProduct_completely_multiplicative_hasProd {f : ℕ →*₀ F} (hsum : Summable (‖f ·‖)) :
+    HasProd (fun p : Primes ↦ (1 - f p)⁻¹) (∑' n, f n) := by
+  have H : (fun p : Primes ↦ (1 - f p)⁻¹) = fun p : Primes ↦ ∑' (e : ℕ), f (p ^ e) :=
+    funext <| fun p ↦ one_sub_inv_eq_geometric_of_summable_norm p.prop hsum
+  simpa only [map_pow, H]
+    using eulerProduct_hasProd f.map_one (fun {m n} _ ↦ f.map_mul m n) hsum f.map_zero
+
+/-- The *Euler Product* for completely multiplicative functions.
+
+If `f : ℕ →*₀ F`, where `F` is a complete normed field and `‖f ·‖` is summable, then
+`∏' p : Nat.Primes, (1 - f p)⁻¹ = ∑' n, f n`. -/
+theorem eulerProduct_completely_multiplicative_tprod {f : ℕ →*₀ F} (hsum : Summable (‖f ·‖)) :
+    ∏' p : Primes, (1 - f p)⁻¹ = ∑' n, f n :=
+  (eulerProduct_completely_multiplicative_hasProd hsum).tprod_eq
 
 open Filter in
 /-- The *Euler Product* for completely multiplicative functions.
-If `f : ℕ →*₀ F`, where `F` is a complete normed field
-and `‖f ·‖` is summable, then `∏' p : {p : ℕ | p.Prime}, (1 - f p)⁻¹ = ∑' n, f n`.
-Since there are no infinite products yet in Mathlib, we state it in the form of
-convergence of finite partial products. -/
--- TODO: Change to use `∏'` once infinite products are in Mathlib
-theorem eulerProduct_completely_multiplicative {f : ℕ →*₀ F} (hsum : Summable fun x => ‖f x‖) :
-    Tendsto (fun n : ℕ ↦ ∏ p in primesBelow n, (1 - f p)⁻¹) atTop (𝓝 (∑' n, f n)) := by
-  convert eulerProduct f.map_one (fun {m n} _ ↦ f.map_mul m n) hsum f.map_zero with N p hN
-  simp_rw [map_pow]
-  refine (tsum_geometric_of_norm_lt_one <| summable_geometric_iff_norm_lt_one.mp ?_).symm
-  refine Summable.of_norm ?_
-  convert hsum.comp_injective <| Nat.pow_right_injective (prime_of_mem_primesBelow hN).one_lt
-  simp only [norm_pow, Function.comp_apply, map_pow]
 
-end CompletelyMultiplicative
+If `f : ℕ →*₀ F`, where `F` is a complete normed field and `‖f ·‖` is summable, then
+`∏' p : Nat.Primes, (1 - f p)⁻¹ = ∑' n, f n`.
+This version is stated in the form of convergence of finite partial products. -/
```

Comment:
This seems annoyingly indirect, in a way which suggests that there is some API missing elsewhere (and since this PR is perhaps the first serious use of the new `HasProd` functionality, this is definitely something we should be looking out for).

The proof I was expecting, somehow, was to lift `primesBelow n` to a `Finset Primes`, and show that it tends to `atTop` wrt the partial order on finsets of primes; then the statement follows from the `HasProd` version by `Tendsto.comp`. What are we missing in order to make this approach work?

Reply:
This is because `HasProd.tendsto_prod_nat` is stated only for functions with domain `ℕ` (this is the same as for the `HasSum` variant, so this is not specific to products).
This should be generalizable to types with an ordering satisfying suitable hypoheses, but right now I have no time to look into this in more detail.